src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -1277,67 +1277,83 @@ private static unsafe int ApplyMask(Span<byte> toMask, int mask, int maskIndex)
-                    if (((long)p % sizeof(int)) == 0)
+                    if (toMaskEnd - toMaskPtr >= Vector<byte>.Count)
                     {
-                        while (count >= sizeof(int))
+                        Vector<byte> maskVector = Vector.AsVectorByte(new Vector<int>(RollRight(mask, maskIndex)));
+
+                        do
                         {
-                            count -= sizeof(int);
-                            *((int*)p) ^= shiftedMask;
-                            p += sizeof(int);
+                            *(Vector<byte>*)toMaskPtr ^= maskVector;
+                            toMaskPtr += Vector<byte>.Count;
                         }
+                        while (toMaskEnd - toMaskPtr >= Vector<byte>.Count);
+                    }
+                }
+
+                // process remaining data (or all, if couldn't use SIMD) 4 bytes at a time.
+
+                if (toMaskEnd - toMaskPtr >= sizeof(int))
+                {
+                    // align our pointer to sizeof(int)
 
-                        // We don't need to update the maskIndex, as its mod-4 value won't have changed.
-                        // `p` points to the remainder.
+                    while ((long)toMaskPtr % sizeof(int) != 0)
+                    {
+                        *toMaskPtr++ ^= maskPtr[maskIndex];
+                        maskIndex = (maskIndex + 1) & 3;
                     }
 
-                    // Process any remaining data a byte at a time.
-                    if (count > 0)
+                    // use int.
+
+                    int rolledMask = RollRight(mask, maskIndex);
+
+                    while (toMaskEnd - toMaskPtr >= sizeof(int))
                     {
-                        byte* maskPtr = (byte*)&mask;
-                        byte* end = p + count;
-                        while (p < end)
-                        {
-                            *p++ ^= maskPtr[maskIndex];
-                            maskIndex = (maskIndex + 1) & 3;
-                        }
+                        *(int*)toMaskPtr ^= rolledMask;
+                        toMaskPtr += sizeof(int);
                     }
                 }
+
+                // do any remaining data a byte at a time.
+
+                while (toMaskPtr != toMaskEnd)
+                {
+                    *toMaskPtr++ ^= maskPtr[maskIndex];
+                    maskIndex = (maskIndex + 1) & 3;
+                }
             }
 
-            // Return the updated index.
             return maskIndex;
         }
 
+        /// <summary>
+        /// Rolls an integer <paramref name="mask"/> to the right by <paramref name="maskIndex"/> bytes. E.g. RollRight(0x11223344, 1) = 0x44112233.
+        /// </summary>
+        private static int RollRight(int mask, int maskIndex)
+        {
+            int maskShift = maskIndex * 8;
+            return (int)(((uint)mask >> maskShift) | ((uint)mask << (32 - maskShift)));
```

Comment:
This seems like something that belongs in the BinaryOperations class

Reply:
https://github.com/dotnet/corefx/blob/0ae011a57253959b79c952219c5fa16389255b7b/src/System.Runtime.Extensions/ref/System.Runtime.Extensions.cs#L1680

Reply:
Thanks, forgot about this :)